sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c

```diff
@@ -0,0 +1,390 @@
+    p2p_ext->addr_numof = 0;
+    uint8_t *tmp = (uint8_t *) (rdo + 1);
+    uint8_t *addr = NULL;
+    uint8_t addr_len = sizeof(ipv6_addr_t) - p2p_ext->compr;
+    uint8_t i = 0;
+    for (i = 0; i < addr_num; i++) {
+        addr = ((uint8_t *) &p2p_ext->addr_vec[i]) + p2p_ext->compr;
+        memcpy(addr, tmp, addr_len);
+        tmp += addr_len;
+        p2p_ext->addr_numof++;
+    }
+
+    if (!p2p_ext->for_me) {
+        ipv6_addr_t *me = NULL;
+        if(gnrc_ipv6_netif_find_by_prefix(&me, &dodag->dodag_id) == KERNEL_PID_UNDEF) {
+            DEBUG("RPL: no address configured\n");
+            return;
+        }
+        addr = ((uint8_t *) &p2p_ext->addr_vec[i]) + p2p_ext->compr;
+        memcpy(addr, ((uint8_t *) me) + p2p_ext->compr, addr_len);
+        p2p_ext->addr_numof++;
+    }
+    else if (p2p_ext->reply) {
+        p2p_ext->stop = true;
+        p2p_ext->dro_ack = true;
+        p2p_ext->dro_delay = GNRC_RPL_P2P_DRO_DELAY;
+    }
+}
+
+static gnrc_pktsnip_t *_build_initial_DRO(gnrc_rpl_p2p_ext_t *p2p_ext)
+{
+    gnrc_pktsnip_t *pkt = NULL, *opt_snip = NULL;
+    gnrc_rpl_p2p_dro_t *dro = NULL;
+    gnrc_rpl_p2p_opt_rdo_t *rdo = NULL;
+    size_t addr_len = (sizeof(ipv6_addr_t) - p2p_ext->compr);
+    uint8_t addr_size = p2p_ext->addr_numof * addr_len;
+
+    for (uint8_t i = p2p_ext->addr_numof; i > 0; i--) {
+        if ((opt_snip = gnrc_pktbuf_add(pkt, NULL, addr_len, GNRC_NETTYPE_UNDEF)) == NULL) {
+            DEBUG("RPL-P2P: cannot add addresses to RDO - no space left in packet buffer\n");
+            gnrc_pktbuf_release(pkt);
+            return NULL;
+        }
+        memcpy(opt_snip->data, &p2p_ext->addr_vec[i-1], addr_len);
+        pkt = opt_snip;
+    }
+
+    if ((opt_snip = gnrc_pktbuf_add(pkt, NULL, sizeof(gnrc_rpl_p2p_opt_rdo_t),
+                                    GNRC_NETTYPE_UNDEF)) == NULL) {
+        DEBUG("RPL-P2P: cannot allocate DRO - no space left in packet buffer\n");
+        gnrc_pktbuf_release(pkt);
+        return NULL;
+    }
+
+    rdo = opt_snip->data;
+    rdo->type = GNRC_RPL_P2P_OPT_RDO;
+    rdo->length = GNRC_RPL_P2P_RDO_LEN + addr_size;
+    rdo->compr_flags = (p2p_ext->hop_by_hop << GNRC_RPL_P2P_RDO_FLAGS_HbH) |
+                       (p2p_ext->compr & GNRC_RPL_P2P_RDO_FLAGS_COMPR);
+    rdo->lmn = (((rdo->length + 2 - sizeof(*rdo)) / addr_len) & GNRC_RPL_P2P_RDO_FLAGS_NEXT_HOP);
+    rdo->target = p2p_ext->target;
+    pkt = opt_snip;
+
+    if ((opt_snip = gnrc_pktbuf_add(pkt, NULL, sizeof(gnrc_rpl_p2p_dro_t),
+                                    GNRC_NETTYPE_UNDEF)) == NULL) {
+        DEBUG("RPL-P2P: cannot allocate RDO - no space left in packet buffer\n");
+        gnrc_pktbuf_release(pkt);
+        return NULL;
+    }
+    dro = opt_snip->data;
+    dro->instance_id = p2p_ext->dodag->instance->id;
+    dro->version_number = 0;
+    dro->flags_rev = byteorder_htons((((p2p_ext->stop << 1) | (p2p_ext->dro_ack << 0))
```

Comment:
yeah that's true, but I like to stress that the flag is at bit position 0 if someone looks at the packet format in the RFC (: